src/essence/cli.py

```diff
@@ -0,0 +1,158 @@
+    generate_spec = args.generate_spec
+    generate_input_template = args.no_template != True
+    functions_to_build = args.functions
+    build_read_none = args.build_read_none
+    build_write_only = args.build_write_only
+
+    if pathlib.Path(input_file).suffix == '.c':
+        print("got .c file as input, you probably meant .bc")
+        return;
+
+    if pathlib.Path(input_file).suffix == '.bc':
+        if build_execs:
+            essence_build(input_file, output_dir, generate_input_template, functions_to_build)
+
+        elif build_read_none or build_write_only:
+            if build_read_none:
+                essence_build_read_none(input_file, output_dir, generate_input_template)
+            if build_write_only:
+                essence_build_write_only(input_file, output_dir, generate_input_template)
+
+        elif generate_spec:
+            essence_generate_spec(input_file, output_dir)
+
+        else:
+            essence_generate_spec(input_file, output_dir)
+            path_to_spec_file = get_filepath_in_output_dir(output_dir, input_file, ".spec.json")
+            essence_list_signatures(path_to_spec_file)
+    else:
+        essence_list_signatures(input_file)
+
+
+#
+#
+#
+# # lists function signatures
+def essence_list_signatures(spec_file: str):
+    with open(spec_file, 'r') as j:
+        contents = json.loads(j.read())
+        funcs = contents['functions']
+
+        funcs.sort(key=lambda content: content['purity'])
+        groups = groupby(funcs, lambda content: content['purity'])
+
+        print("-------- Functions --------")
+        for purity, funcs in groups:
+            print(purity + ":")
+            for func in funcs:
+                print("\t", func['name'] + ":", func['signature'])
+
+
+#
+# # generates and prints spec
+def essence_generate_spec(bc_file: str, output: str):
+    subprocess.run([handsan_path, "-o", output, bc_file])
+
+
+#
+#
+#
+# # build the actual functions
+def essence_build(bc_file: str, output: str, generate_input_template: bool, function_names: [str]):
+    print("----------------- building functions --------------------")
+    for func_name in function_names:
+        print(f"building: {func_name}")
+        build_functions_for(bc_file, output, generate_input_template, func_name)
+
+
+
+def essence_build_read_none(bc_file: str, output: str, generate_input_template: bool):
+    essence_build_for_purity_level(bc_file, output, generate_input_template, 'read_none')
+
+def essence_build_write_only(bc_file: str, output: str, generate_input_template: bool):
+    essence_build_for_purity_level(bc_file, output, generate_input_template, 'write_only')
+
+
+def essence_build_for_purity_level(bc_file: str, output: str, generate_input_template: bool, purity_level :str):
+    essence_generate_spec(bc_file, output)
+    spec_path = get_filepath_in_output_dir(output, bc_file, '.spec.json')
+    with open(spec_path, 'r') as j:
+        contents = json.loads(j.read())
```

Comment:
Use the direct IO form:

```suggestion
        contents = json.load(j)
```